{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import sys \n",
    "import commot as ct\n",
    "\n",
    "sys.path.append('../../src')\n",
    "sys.path.append('..')\n",
    "\n",
    "from spaceoracle.tools.network import expand_paired_interactions\n",
    "from spaceoracle.tools.network import get_mouse_housekeeping_genes\n",
    "from spaceoracle.oracles import BaseTravLR\n",
    "from spaceoracle.tools.network import encode_labels\n",
    "\n",
    "import celloracle as co"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 16280 × 24649\n",
       "    obs: 'cluster', 'spot_class', 'first_type', 'second_type', 'first_class', 'second_class', 'min_score', 'singlet_score', 'conv_all', 'conv_doublet', '1', '2', '3', '4', '5', '6', '7'\n",
       "    uns: 'cluster_colors'\n",
       "    obsm: 'X_spatial', 'spatial'"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/Users/koush/Projects/SpaceOracle/data/slideseq/mLND3_1.h5ad')\n",
    "adata.obs = adata.obs.join(\n",
    "    pd.read_csv('/Users/koush/Projects/SpaceOracle/data/slideseq/day3_1_rctd.csv', index_col=0))\n",
    "adata = adata[adata.obs.spot_class.isin(['singlet', 'doublet_certain'])]\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading prebuilt promoter base-GRN. Version: mm10_gimmemotifsv5_fpr2\n"
     ]
    }
   ],
   "source": [
    "base_GRN = co.data.load_mouse_promoter_base_GRN()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ligrec = ct.pp.ligand_receptor_database(\n",
    "        database='CellChat', \n",
    "        species='mouse', \n",
    "        signaling_type=\"Secreted Signaling\"\n",
    "    )\n",
    "            \n",
    "df_ligrec.columns = ['ligand', 'receptor', 'pathway', 'signaling']\n",
    "df_ligrec = expand_paired_interactions(df_ligrec)\n",
    "\n",
    "coms_df =df_ligrec.query('receptor.str.contains(\"Cxcr\") or ligand.str.contains(\"Cxcr\")')\n",
    "genes_of_interest = list(np.unique(coms_df.ligand.unique().tolist() + \\\n",
    "                         coms_df.receptor.unique().tolist()))\n",
    "genes_of_interest = genes_of_interest\n",
    "len(genes_of_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilr2_df =df_ligrec.query('receptor.str.contains(\"Il2\") or ligand.str.contains(\"Il2\")')\n",
    "genes_of_interest = list(np.unique(ilr2_df.ligand.unique().tolist() + ilr2_df.receptor.unique().tolist()))\n",
    "# len(genes_of_interest)\n",
    "genes_of_interest = genes_of_interest + ['Pax5', 'Bach2', 'Prdm1', 'Bcl6', 'Foxp3', 'Satb1', 'Id2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "housekeeping_db = get_mouse_housekeeping_genes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_top_genes = 4000\n",
    "min_cells = 10\n",
    "min_counts = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chromosome_scaffold</th>\n",
       "      <th>Transcript start</th>\n",
       "      <th>Transcript end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000007921</td>\n",
       "      <td>0610009B22Rik</td>\n",
       "      <td>11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000077783</td>\n",
       "      <td>0610030E20Rik</td>\n",
       "      <td>6</td>\n",
       "      <td>72347334</td>\n",
       "      <td>72353148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000192144</td>\n",
       "      <td>0610030E20Rik</td>\n",
       "      <td>CHR_MG184_PATCH</td>\n",
       "      <td>72347333</td>\n",
       "      <td>72353147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000125107</td>\n",
       "      <td>0610037L13Rik</td>\n",
       "      <td>4</td>\n",
       "      <td>107889899</td>\n",
       "      <td>107897802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000106727</td>\n",
       "      <td>0610037L13Rik</td>\n",
       "      <td>4</td>\n",
       "      <td>107889863</td>\n",
       "      <td>107897118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3979</th>\n",
       "      <td>ENSMUST00000101388</td>\n",
       "      <td>Zxdb</td>\n",
       "      <td>X</td>\n",
       "      <td>94724569</td>\n",
       "      <td>94730187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3980</th>\n",
       "      <td>ENSMUST00000113539</td>\n",
       "      <td>Zxdc</td>\n",
       "      <td>6</td>\n",
       "      <td>90369492</td>\n",
       "      <td>90385084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3981</th>\n",
       "      <td>ENSMUST00000203493</td>\n",
       "      <td>Zxdc</td>\n",
       "      <td>6</td>\n",
       "      <td>90369613</td>\n",
       "      <td>90383103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3982</th>\n",
       "      <td>ENSMUST00000075117</td>\n",
       "      <td>Zxdc</td>\n",
       "      <td>6</td>\n",
       "      <td>90369494</td>\n",
       "      <td>90403490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3983</th>\n",
       "      <td>ENSMUST00000043616</td>\n",
       "      <td>Zyg11b</td>\n",
       "      <td>4</td>\n",
       "      <td>108229724</td>\n",
       "      <td>108301096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3984 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Ensembl ID           Gene Chromosome_scaffold  Transcript start  \\\n",
       "0     ENSMUST00000007921  0610009B22Rik                  11          51685386   \n",
       "1     ENSMUST00000077783  0610030E20Rik                   6          72347334   \n",
       "2     ENSMUST00000192144  0610030E20Rik     CHR_MG184_PATCH          72347333   \n",
       "3     ENSMUST00000125107  0610037L13Rik                   4         107889899   \n",
       "4     ENSMUST00000106727  0610037L13Rik                   4         107889863   \n",
       "...                  ...            ...                 ...               ...   \n",
       "3979  ENSMUST00000101388           Zxdb                   X          94724569   \n",
       "3980  ENSMUST00000113539           Zxdc                   6          90369492   \n",
       "3981  ENSMUST00000203493           Zxdc                   6          90369613   \n",
       "3982  ENSMUST00000075117           Zxdc                   6          90369494   \n",
       "3983  ENSMUST00000043616         Zyg11b                   4         108229724   \n",
       "\n",
       "      Transcript end  \n",
       "0           51688653  \n",
       "1           72353148  \n",
       "2           72353147  \n",
       "3          107897802  \n",
       "4          107897118  \n",
       "...              ...  \n",
       "3979        94730187  \n",
       "3980        90385084  \n",
       "3981        90383103  \n",
       "3982        90403490  \n",
       "3983       108301096  \n",
       "\n",
       "[3984 rows x 5 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "housekeeping_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16136 × 4016\n",
       "    obs: 'cluster', 'spot_class', 'first_type', 'second_type', 'first_class', 'second_class', 'min_score', 'singlet_score', 'conv_all', 'conv_doublet', '1', '2', '3', '4', '5', '6', '7', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'cluster_colors', 'log1p', 'hvg'\n",
       "    obsm: 'X_spatial', 'spatial'\n",
       "    layers: 'raw_count', 'normalized_count'"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names_make_unique()\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], inplace=True)\n",
    "sc.pp.filter_cells(adata, min_counts=min_counts)\n",
    "adata = adata[adata.obs[\"pct_counts_mt\"] < 20].copy()\n",
    "adata = adata[:, ~adata.var[\"mt\"]]\n",
    "adata = adata[:, ~adata.var_names.isin(housekeeping_db['Gene'].unique())]\n",
    "\n",
    "sc.pp.filter_genes(adata, min_cells=min_cells)\n",
    "adata.layers[\"raw_count\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata, flavor=\"seurat\", n_top_genes=n_top_genes)\n",
    "\n",
    "adata = adata[:, (adata.var.highly_variable | adata.var_names.isin(genes_of_interest))]\n",
    "\n",
    "adata.layers[\"normalized_count\"] = adata.to_df().values.copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in ['n_genes_by_counts', 'log1p_n_genes_by_counts',\n",
    "       'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes',\n",
    "       'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes',\n",
    "       'pct_counts_in_top_500_genes', 'total_counts_MT',\n",
    "       'log1p_total_counts_MT', 'pct_counts_MT', 'n_counts']:\n",
    "    if key in adata.obs.keys():\n",
    "        adata.obs.pop(key)\n",
    "\n",
    "for key in ['MT', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts',\n",
    "       'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts',\n",
    "       'n_cells', 'highly_variable', 'means', 'dispersions',\n",
    "       'dispersions_norm']:\n",
    "    if key in adata.var.keys():\n",
    "        adata.var.pop(key)\n",
    "\n",
    "del adata.uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype('category')\n",
    "\n",
    "adata.obs['cell_type_int'] = adata.obs['cell_type'].apply(\n",
    "    lambda x: encode_labels(adata.obs['cell_type'], reverse_dict=True)[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5778 × 3540\n",
       "    obs: 'cell_type', 'cell_type_int'\n",
       "    obsm: 'spatial'\n",
       "    layers: 'normalized_count'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating MAGIC...\n",
      "  Running MAGIC on 5778 cells and 3540 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 10.15 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 1.93 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 1.62 seconds.\n",
      "  Calculated graph and diffusion operator in 13.73 seconds.\n",
      "  Calculating imputation...\n",
      "  Calculated imputation in 0.55 seconds.\n",
      "Calculated MAGIC in 14.30 seconds.\n"
     ]
    }
   ],
   "source": [
    "pcs = BaseTravLR.perform_PCA(adata)\n",
    "BaseTravLR.knn_imputation(adata, pcs, method='MAGIC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5778 × 3540\n",
       "    obs: 'cell_type', 'cell_type_int'\n",
       "    obsm: 'spatial'\n",
       "    layers: 'normalized_count', 'imputed_count'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5778, 3540)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers['imputed_count'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('/Users/koush/Desktop/training_data/snrna_human_tonsil.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bee",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
